File: Android/app/src/main/java/com/example/supplementtracker/domain/util/DoseTimingPolicy.kt
package com.example.supplementtracker.domain.util

import com.example.supplementtracker.domain.model.IntakeStatus

/** Single source of truth for dose-state timing and adherence formulas. */
object DoseTimingPolicy {
    const val SOON_WINDOW_MS: Long = 20L * 60L * 1000L
    const val MISSED_AFTER_MS: Long = 2L * 60L * 60L * 1000L

    fun isDueSoon(scheduledAtEpochMs: Long, nowEpochMs: Long): Boolean =
        scheduledAtEpochMs > nowEpochMs && scheduledAtEpochMs - nowEpochMs <= SOON_WINDOW_MS

    fun isMissed(scheduledAtEpochMs: Long, nowEpochMs: Long): Boolean =
        scheduledAtEpochMs > 0L && nowEpochMs > scheduledAtEpochMs + MISSED_AFTER_MS

    fun isMissedSoon(scheduledAtEpochMs: Long, nowEpochMs: Long): Boolean {
        val missedAt = scheduledAtEpochMs + MISSED_AFTER_MS
        return scheduledAtEpochMs > 0L && nowEpochMs in (missedAt - SOON_WINDOW_MS) until missedAt
    }

    fun isLateTaken(status: String?, scheduledAtEpochMs: Long, updatedAtEpochMs: Long): Boolean =
        status == IntakeStatus.TAKEN.storageValue && updatedAtEpochMs > 0L &&
            updatedAtEpochMs > scheduledAtEpochMs + SOON_WINDOW_MS

    fun completionRate(taken: Int, skipped: Int): Double? {
        val total = taken + skipped
        return if (total <= 0) null else taken.toDouble() / total.toDouble()
    }

    fun completionPercent(taken: Int, skipped: Int): Int? =
        completionRate(taken, skipped)?.let { (it * 100.0).toInt() }
}

File: iOS/Services/DoseTimingPolicy.swift
import Foundation

/// Canonical dose timing and adherence formulas shared by Home, History and reports.
public enum DoseTimingPolicy: Sendable {
    public static let soonWindowMilliseconds: Int64 = 20 * 60 * 1_000
    public static let missedAfterMilliseconds: Int64 = 2 * 60 * 60 * 1_000

    public static func isDueSoon(scheduledAtEpochMs: Int64, nowEpochMs: Int64) -> Bool {
        scheduledAtEpochMs > nowEpochMs && scheduledAtEpochMs - nowEpochMs <= soonWindowMilliseconds
    }

    public static func isMissed(scheduledAtEpochMs: Int64, nowEpochMs: Int64) -> Bool {
        scheduledAtEpochMs > 0 && nowEpochMs > scheduledAtEpochMs + missedAfterMilliseconds
    }

    public static func isMissedSoon(scheduledAtEpochMs: Int64, nowEpochMs: Int64) -> Bool {
        let missedAt = scheduledAtEpochMs + missedAfterMilliseconds
        return scheduledAtEpochMs > 0 && nowEpochMs >= missedAt - soonWindowMilliseconds && nowEpochMs < missedAt
    }

    public static func isLateTaken(status: String, scheduledAtEpochMs: Int64, updatedAtEpochMs: Int64) -> Bool {
        status == IntakeStatus.taken.rawValue && updatedAtEpochMs > 0 &&
            updatedAtEpochMs > scheduledAtEpochMs + soonWindowMilliseconds
    }

    public static func completionRate(taken: Int, skipped: Int) -> Double? {
        let total = taken + skipped
        guard total > 0 else { return nil }
        return Double(taken) / Double(total)
    }

    public static func completionPercent(taken: Int, skipped: Int) -> Int? {
        completionRate(taken: taken, skipped: skipped).map { Int($0 * 100) }
    }
}
